py2puml/parsing/astvisitors.py

```diff
@@ -0,0 +1,208 @@
+        if self.class_self_id is None:
+            self.class_self_id = variable.id
+        # other arguments are constructor parameters
+        else:
+            self.variables.append(variable)
+
+
+class AssignedVariablesCollector(NodeVisitor):
+    '''Parses the target of an assignment statement to detect whether the value is assigned to a variable or an instance attribute'''
+    def __init__(self, class_self_id: str, annotation: expr):
+        self.class_self_id: str = class_self_id
+        self.annotation: expr = annotation
+        self.variables: List[Variable] = []
+        self.self_attributes: List[Variable] = []
+
+    def visit_Name(self, node: Name):
+        '''
+        Detects declarations of new variables
+        '''
+        if node.id != self.class_self_id:
+            self.variables.append(Variable(node.id, self.annotation))
+
+    def visit_Attribute(self, node: Attribute):
+        '''
+        Detects declarations of new attributes on 'self'
+        '''
+        if isinstance(node.value, Name) and node.value.id == self.class_self_id:
+            self.self_attributes.append(Variable(node.attr, self.annotation))
+
+    def visit_Subscript(self, node: Subscript):
+        '''
+        Assigns a value to a subscript of an existing variable: must be skipped
+        '''
+        pass
+
+
+class ConstructorVisitor(NodeVisitor):
+    '''Identifies the attributes (and infer their type) assigned to self in the body of a constructor method'''
+    def __init__(self, constructor_source: str, class_name: str, root_fqdn: str, module_resolver: ModuleResolver, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.constructor_source = constructor_source
+        self.class_fqdn: str = f'{module_resolver.module.__name__}.{class_name}'
+        self.root_fqdn = root_fqdn
+        self.module_resolver = module_resolver
+        self.class_self_id: str = None
+        self.variables_namespace: List[Variable] = []
+        self.uml_attributes: List[UmlAttribute] = []
+        self.uml_relations_by_target_fqdn: Dict[str, UmlRelation] = {}
+
+    def extend_relations(self, target_fqdns: List[str]):
+        self.uml_relations_by_target_fqdn.update({
+            target_fqdn: UmlRelation(self.class_fqdn, target_fqdn, RelType.COMPOSITION)
+            for target_fqdn in target_fqdns
+            if target_fqdn.startswith(self.root_fqdn) and target_fqdn not in self.uml_relations_by_target_fqdn
+        })
+
+    def is_in_variables_namespace(self, variable_id: str) -> bool:
+        return self.get_from_namespace(variable_id) != None
+
+    def get_from_namespace(self, variable_id: str) -> Variable:
+        return next((
+            variable
+            # variables namespace is iterated antichronologically
+            # to account for variables being overridden
+            for variable in self.variables_namespace[::-1]
+            if variable.id == variable_id
+        ), None)
```

Comment:
```suggestion
    def get_from_namespace(self, variable_id: str) -> Optional[Variable]:
        return next((
            variable
            # variables namespace is iterated antichronologically
            # to account for variables being overridden
            for variable in self.variables_namespace[::-1]
            if variable.id == variable_id
        ), None)
```

Reply:
agreed. There used to be several calls at some point, but there was just one now. Thanks